cluster-autoscaler/processors/nodeinfos/refine_node_infos_processor.go

```diff
@@ -0,0 +1,174 @@
+	for _, nodeGroup := range ctx.CloudProvider.NodeGroups() {
+		nodeGroups[nodeGroup.Id()] = nodeGroup
+	}
+
+	if err := p.nodeInfoCache.update(nodeGroups); err != nil {
+		return nodeInfosForNodeGroups, err
+	}
+	p.nodeInfoCache.RLock()
+	defer p.nodeInfoCache.RUnlock()
+
+	ignoredTaints := make(taints.TaintKeySet)
+	for _, taintKey := range ctx.IgnoredTaints {
+		ignoredTaints[taintKey] = true
+	}
+
+	// build comparable (all from templates) NodeInfos
+	for groupID, nodeInfo := range nodeInfosForNodeGroups {
+		templatedNodeInfo, ok := p.nodeInfoCache.nodeInfos[groupID]
+		if !ok || templatedNodeInfo == nil {
+			continue
+		}
+
+		templatesNodeInfo[groupID] = templatedNodeInfo
+
+		if utils.IsNodeInfoBuiltFromTemplate(nodeInfo) {
+			innacurateNodeInfos[groupID] = templatedNodeInfo
+		}
+	}
+
+	// refine innacurate nodeinfos when we can find similar nodeInfos built from real nodes
+	for groupID, nodeInfo := range innacurateNodeInfos {
+		nodeGroup, ok := nodeGroups[groupID]
+		if !ok {
+			continue
+		}
+
+		similars, err := p.NodeGroupSetProcessor.FindSimilarNodeGroups(ctx, nodeGroup, templatesNodeInfo)
+		if err != nil {
+			klog.Warningf("Failed to lookup for matching node groups for %s: %v", groupID, err)
+			return nodeInfosForNodeGroups, nil
+		}
+
+		for _, nodeGroup := range similars {
+			similarInfo, found := nodeInfosForNodeGroups[nodeGroup.Id()]
+			if !found || utils.IsNodeInfoBuiltFromTemplate(similarInfo) {
+				continue
+			}
+			refinedNodeInfo, err := mirrorNodeInfoFromSimilarGroup(nodeInfo, similarInfo, groupID, ignoredTaints)
+			if err != nil {
+				return nodeInfosForNodeGroups, err
+			}
+			nodeInfosForNodeGroups[groupID] = refinedNodeInfo
+			break
+		}
+	}
+
+	return nodeInfosForNodeGroups, nil
+}
+
+// CleanUp cleans up processor's internal structures.
+func (p *RefineNodeInfosProcessor) CleanUp() {
+}
+
+func mirrorNodeInfoFromSimilarGroup(oldInfo, newInfo *schedulerframework.NodeInfo, nodeGroupName string, ignoredTaints taints.TaintKeySet) (*schedulerframework.NodeInfo, errors.AutoscalerError) {
+	newInfoCopy := newInfo.Clone()
+
+	// keep locality labels (region, zone, failure domain, ...) from original template
+	newNode := newInfoCopy.Node().DeepCopy()
+	for label, val := range oldInfo.Node().GetLabels() {
+		if _, found := nodegroupset.BasicIgnoredLabels[label]; found {
+			newNode.Labels[label] = val
+		}
+	}
+	newInfoCopy.SetNode(newNode)
+	utils.SetNodeInfoBuiltFromTemplate(newInfoCopy)
+
+	// deep copy pods and set node name (in node, pods, labels...) to the nodegroup's
+	return utils.SanitizeNodeInfo(newInfoCopy, nodeGroupName, ignoredTaints)
+}
+
+func (n *nodeInfoCache) update(nodeGroups map[string]cloudprovider.NodeGroup) error {
+	n.Lock()
+	defer n.Unlock()
+
+	if len(n.nodeInfos) == 0 {
+		n.nodeInfos = make(map[string]*schedulerframework.NodeInfo)
+		n.lastUpdated = time.Now()
+	}
+	needsRefresh := n.lastUpdated.Add(templateNodeInfoCacheTTL).Before(time.Now())
```

Comment:
I think we could remove cache TTL. Instead we could,
1. If the new nodeinfo is real node, update the cache.
2. Else, use cached real nodeinfo.

Assume we never or  rarely change instance type within a nodegroup. So in most time, the cache works well.
Anyway if we do change , our cache will be updated later(when scaling up from 1 or more).

Reply:
Not sure to understand the suggestion. We don't cache nodeInfos from real-world nodes, only templated nodeInfos which we can then use as keys for comparisons/findsimilar (then once similar groups ids are found, we use real-world nodeinfos as provided as an arg to that processor (no need to cache those)).

Cache TTL is a distinct matter; it's not essential and can be removed from the current implememtation if we prefer so; it's meant to eventually catch up if a nodegroup is updated (or deleted+recreated): scale-up might never happen with stale infos.